File: Back/NodeJS/modules/bets/src/domain/members/Member.ts
import { MemberId } from './MemberId'
import { Friendship } from '../friendships/Friendship';
import { Bet } from '../bets/Bet';
import { BetId } from '../bets/BetId';
import { NotEnoughChipsException } from './exceptions/NotEnoughChipsException';
import { NoneFriendException } from './exceptions/NoneFriendException';
import { AnswerBet } from '../answerBets/AnswerBet';
import { IDateTimeProvider } from '../../../../shared/domain/IDateTimeProvider';
import { MemberSnapshot } from './MemberSnapshot';

export class Member {
    static FromSnapshot(snapshot: MemberSnapshot): Member {
        return new Member(new MemberId(snapshot.MemberId), snapshot.Username, snapshot.Coins, snapshot.CountFriends)
    }
    
    static Create(memberId: string, username: string): Member {
        return new Member(new MemberId(memberId), username, 2000, 0)
    }

    private constructor(private readonly memberId: MemberId, 
                private readonly username: string,
                private coins: number, 
                private countFriend: number){}

    Bet(betId: string, 
        description: string, 
        coins: number, 
        endDate: Date, 
        members: string[],
        dateTimeProvider: IDateTimeProvider): Bet {
        if(this.coins < coins) {
            throw new NotEnoughChipsException();
        }

        if(this.countFriend == 0) {
            throw new NoneFriendException();
        }

        return Bet.Create(new BetId(betId),
                        this.memberId,
                        description,
                        coins,
                        endDate,
                        members,
                        dateTimeProvider);
    }

    DecreaseBalance(coins: number) {
        this.coins -= coins; 
    }

    IncreaseBalance(coins: number) {
        this.coins += coins
    }
    
    CanBet(bet: Bet): boolean {
        if(this.coins < bet.Coins) {
            return false;
        }
        return true;
    }

    AddFriendship(requesterId: string) {
        return Friendship.Create(new MemberId(requesterId), this.memberId);
    }
    
    RejectBet(betId: BetId) {
        return new AnswerBet(betId, false, this.memberId);
    }

    AcceptBet(betId: BetId) {
        return new AnswerBet(betId, true, this.memberId);
    }

    GetSnapshot(): MemberSnapshot {
        return new MemberSnapshot(this.memberId.Value,
                                this.username,
                                this.coins,
                                this.countFriend)
    }
}

File: Back/.Net/Modules/Bets/BetFriends.Bets.Domain/Members/Member.cs
using BetFriends.Bets.Domain.Friendships;
using BetFriends.Bets.Domain.Members.Exceptions;
using BetFriends.Bets.Domain.Bets;
using BetFriends.Shared.Domain;

namespace BetFriends.Bets.Domain.Members;

public class Member : Entity
{
    private readonly MemberId memberId;
    private readonly string username;
    private int coins;
    private readonly int countFriends;

    private Member(MemberId memberId, string username, int coins)
    {
        this.memberId = memberId;
        this.username = username;
        this.coins = coins;
    }

    private Member(MemberState memberState)
    {
        memberId = new MemberId(memberState.MemberId);
        username = memberState.Username;
        coins = memberState.Coins;
        countFriends = memberState.CountFriends;
    }

    public MemberId MemberId { get => memberId; }
    public MemberState Snapshot { get => new(memberId.Value, username, coins, countFriends); }

    public Friendship AddFriendship(Guid requesterId)
    {
        return Friendship.Create(memberId.Value, requesterId);
    }

    public Bet Bet(BetId betId,
                   string description,
                   int coins,
                   DateTime endDate,
                   IEnumerable<Guid> friends,
                   DateTime creationDate)
    {
        if (this.coins < coins)
            throw new CannotBetException(CannotBetException.CoinsNotEnough);

        if (!HasFriend())
            throw new CannotBetException(CannotBetException.NoneFriends);
        return Bets.Bet.Create(betId,
                               memberId,
                               description,
                               coins,
                               endDate,
                               friends,
                               creationDate);
    }

    public void DecreaseBalance(int coins)
    {
        this.coins -= coins;
    }

    private bool HasFriend()
    {
        return countFriends != 0;
    }

    internal bool CanBet(int coins)
        => this.coins >= coins;

    public void IncreaseBalance(int coins)
        => this.coins += coins;

    public static Member Create(MemberId memberId, string username)
    {
        return new Member(memberId, username, 2000);
    }

    public static Member FromState(MemberState memberState)
    {
        return new Member(memberState);
    }
}
